api/types/instance.go

```diff
@@ -0,0 +1,256 @@
+	}
+}
+
+func (i *InstanceV1) SyncLogAndResourceExpiry(ttl time.Duration) {
+	// expire control log entries relative to LastSeen.
+	logExpiry := i.expireControlLog(i.Spec.LastSeen, ttl)
+
+	// calculate the default resource expiry.
+	resourceExpiry := i.Spec.LastSeen.Add(ttl)
+
+	// if one or more log entries want to outlive the default resource
+	// expiry, we bump the resource expiry to match.
+	if logExpiry.After(resourceExpiry) {
+		resourceExpiry = logExpiry
+	}
+
+	i.Metadata.SetExpiry(resourceExpiry.UTC())
+}
+
+func (i *InstanceV1) GetTeleportVersion() string {
+	return i.Spec.Version
+}
+
+func (i *InstanceV1) GetServices() []SystemRole {
+	return i.Spec.Services
+}
+
+func (i *InstanceV1) HasService(s SystemRole) bool {
+	return slices.Contains(i.Spec.Services, s)
+}
+
+func (i *InstanceV1) GetHostname() string {
+	return i.Spec.Hostname
+}
+
+func (i *InstanceV1) GetAuthID() string {
+	return i.Spec.AuthID
+}
+
+func (i *InstanceV1) GetLastSeen() time.Time {
+	return i.Spec.LastSeen
+}
+
+func (i *InstanceV1) SetLastSeen(t time.Time) {
+	i.Spec.LastSeen = t.UTC()
+}
+
+func (i *InstanceV1) GetControlLog() []InstanceControlLogEntry {
+	return i.Spec.ControlLog
+}
+
+func (i *InstanceV1) AppendControlLog(entries ...InstanceControlLogEntry) {
+	n := len(i.Spec.ControlLog)
+	i.Spec.ControlLog = append(i.Spec.ControlLog, entries...)
+	for idx, entry := range i.Spec.ControlLog[n:] {
+		// ensure that all provided timestamps are UTC (non-UTC timestamps can cause
+		// panics in proto logic).
+		i.Spec.ControlLog[idx].Time = entry.Time.UTC()
+	}
+	slices.SortFunc(i.Spec.ControlLog, func(a, b InstanceControlLogEntry) bool {
+		return a.Time.Before(b.Time)
+	})
+}
+
+// expireControlLog removes expired entries from the control log relative to the supplied
+// "now" value. The supplied ttl is used as the default ttl for entries that do not specify
+// a custom ttl value. The returned timestamp is the observed expiry that was furthest in
+// the future.
+func (i *InstanceV1) expireControlLog(now time.Time, ttl time.Duration) time.Time {
+	now = now.UTC()
+	filtered := i.Spec.ControlLog[:0]
+	var latestExpiry time.Time
+	for _, entry := range i.Spec.ControlLog {
+		entryTTL := entry.TTL
+		if entryTTL == 0 {
+			entryTTL = ttl
+		}
+		if entry.Time.IsZero() {
+			entry.Time = now
+		}
+		expiry := entry.Time.Add(entryTTL)
+		if now.After(expiry) {
+			continue
+		}
+
+		if expiry.After(latestExpiry) {
+			latestExpiry = expiry
+		}
+		filtered = append(filtered, entry)
+	}
+	// ensure that we don't preserve pointers in the now out of
+	// range portion of the control log by zeroing the diff.
+	for idx := len(filtered); idx < len(i.Spec.ControlLog); idx++ {
+		i.Spec.ControlLog[idx] = InstanceControlLogEntry{}
+	}
+	i.Spec.ControlLog = filtered
+	return latestExpiry
+}
+
+func (i *InstanceV1) Clone() Instance {
+	return proto.Clone(i).(*InstanceV1)
+}
+
+func (e *InstanceControlLogEntry) Clone() InstanceControlLogEntry {
+	e.Time = e.Time.UTC()
```

Comment:
This would be a write even if the written value is identical to what's already there, I think - which would violate the expectation that `Clone()` is a read-only operation.

Perhaps something like this?
```suggestion
	if e.Time != e.Time.UTC() {
		shallow := *e
		e = &shallow
		e.Time = e.Time.UTC()
	}
```

Reply:
(we've had several instances of `CheckAndSetDefaults` calls causing data races in the past, this could end up doing a similar thing)